x/ccv/provider/keeper/punish_validator.go

```diff
@@ -31,8 +31,76 @@ func (k Keeper) JailValidator(ctx sdk.Context, providerAddr types.ProviderConsAd
 		k.stakingKeeper.Jail(ctx, providerAddr.ToSdkConsAddr())
 	}
 
-	// update jail time to end after double sign jail duration
+	// Jail the validator to trigger the unbonding of the validator
+	// (see cosmos/cosmos-sdk/blob/v0.45.16-ics-lsm/x/staking/keeper/val_state_change.go#L192).
 	k.slashingKeeper.JailUntil(ctx, providerAddr.ToSdkConsAddr(), evidencetypes.DoubleSignJailEndTime)
 
-	// TODO: add tombstoning back once we integrate the slashing
+	// Tombstone the validator so that we cannot slash the validator more than once
+	// (see cosmos/cosmos-sdk/blob/v0.45.16-ics-lsm/x/evidence/keeper/infraction.go#L81).
+	// Note that we cannot simply use the fact that a validator is jailed to avoid slashing more than once
+	// because then a validator could i) perform an equivocation, ii) get jailed (e.g., through downtime)
+	// and in such a case the validator would not get slashed when we call `SlashValidator`.
+	k.slashingKeeper.Tombstone(ctx, providerAddr.ToSdkConsAddr())
+}
+
+// ComputePowerToSlash computes the power to be slashed based on the tokens in non-matured (based on the
+// provider `now` time) `undelegations` and `redelegations`, as well as the current `power` of the validator
+func (k Keeper) ComputePowerToSlash(now time.Time, undelegations []stakingtypes.UnbondingDelegation,
+	redelegations []stakingtypes.Redelegation, power int64, powerReduction sdk.Int) int64 {
+
+	// compute the total numbers of tokens currently being undelegated
+	undelegationsInTokens := sdk.NewInt(0)
+	for _, u := range undelegations {
+		for _, entry := range u.Entries {
+			if entry.IsMature(now) && !entry.OnHold() {
+				//  undelegation no longer eligible for slashing, skip it
+				continue
+			}
+			undelegationsInTokens = undelegationsInTokens.Add(entry.InitialBalance)
+		}
+	}
+
+	// compute the total numbers of tokens currently being redelegated
+	redelegationsInTokens := sdk.NewInt(0)
+	for _, r := range redelegations {
+		for _, entry := range r.Entries {
+			if entry.IsMature(now) && !entry.OnHold() {
+				//  redelegation no longer eligible for slashing, skip it
+				continue
+			}
+			redelegationsInTokens = redelegationsInTokens.Add(entry.InitialBalance)
+		}
+	}
+
+	// The power we pass to staking's keeper `Slash` method is the current power of the validator together with the total
+	// power of all the currently undelegated and redelegated tokens (see docs/docs/adrs/adr-013-equivocation-slashing.md).
+	undelegationsAndRedelegationsInPower := sdk.TokensToConsensusPower(
+		undelegationsInTokens.Add(redelegationsInTokens), powerReduction)
+
+	return power + undelegationsAndRedelegationsInPower
+}
+
+// SlashValidator slashes validator with `providerAddr`
+func (k Keeper) SlashValidator(ctx sdk.Context, providerAddr types.ProviderConsAddress) {
+	logger := k.Logger(ctx)
+
+	val, found := k.stakingKeeper.GetValidatorByConsAddr(ctx, providerAddr.ToSdkConsAddr())
+	if !found {
```

Comment:
Why the case where the validator is unbonded isn't handled?
In this case, the slash method will panic, see:
 see https://github.com/cosmos/cosmos-sdk/blob/0af2f4da004cbea6414a8bad56e8bdcd45badf1e/x/staking/keeper/slash.go#L61-L63C1

Reply:
Nice catch! Fixed it and added a test. Thanks.